pulp_container/app/tasks/builder.py

```diff
@@ -0,0 +1,132 @@
+from pulp_container.app.models import (
+    Blob,
+    BlobManifest,
+    ContainerRepository,
+    Manifest,
+    Tag
+)
+from pulp_container.constants import MEDIA_TYPE
+from pulpcore.app.models.content import Artifact, ContentArtifact
+from pulpcore.plugin.tasking import WorkingDirectory
+
+import json
+import os
+import shutil
+import subprocess
+from uuid import uuid4
+
+
+def get_or_create_blob(layer_json, manifest, path):
+    """
+    Creates Blob from json snippet of manifest.json
+
+    Args:
+        layer_json (json): json
+        path (str): Path of the directory that contains layer
+
+    Returns:
+        class:`pulp_container.app.models.Blob`
+
+    """
+    try:
+        return Blob.objects.get(digest=layer_json['digest'])
+    except Blob.DoesNotExist:
+        pass
+    layer_file_name = "{}{}".format(path, layer_json['digest'][7:])
+    layer_artifact = Artifact.init_and_validate(layer_file_name)
+    layer_artifact.save()
+    blob = Blob(digest=layer_json['digest'], media_type=layer_json['mediaType'])
+    blob.save()
+    ContentArtifact(artifact=layer_artifact, content=blob,
+                    relative_path=layer_json['digest']).save()
+    BlobManifest(manifest=manifest, manifest_blob=blob).save()
+    return blob
+
+
+def add_image_from_directory_to_repository(path, repository, tag):
+    """
+    Creates a Manifest and all blobs from a directory with OCI image
+
+    Args:
+        path (str): Path to directory with the OCI image
+        repository (class:`pulpcore.plugin.models.Repository`): The destination repository
+        tag (str): Tag name for the new image in the repository
+
+    Returns:
+        A class:`pulpcore.plugin.models.RepositoryVersion` that contains the new OCI container
+        image and tag.
+
+    """
+    manifest_path = "{}manifest.json".format(path)
+    manifest_artifact = Artifact.init_and_validate(manifest_path)
+    manifest_artifact.save()
+    manifest_digest = "sha256:{}".format(manifest_artifact.sha256)
+    manifest = Manifest(digest=manifest_digest, schema_version=2,
+                        media_type=MEDIA_TYPE.MANIFEST_OCI)
+    manifest.save()
+    ContentArtifact(artifact=manifest_artifact, content=manifest,
+                    relative_path=manifest_digest).save()
+    latest_tag = Tag(name=tag, tagged_manifest=manifest)
+    latest_tag.save()
+    ContentArtifact(artifact=manifest_artifact, content=latest_tag, relative_path='latest').save()
```

Comment:
s/'latest'/tag

Reply:
thank you! will rename.

Reply:
sorry, i needed to be more explicit, we need to use the name of the tag for the relative_path, and right now it is hard coded to 'latest'
so it should be `relative_path=tag.name`